Core/include/Acts/SpacePointFormation/detail/SingleHitSpacePointBuilder.ipp

```diff
@@ -0,0 +1,100 @@
+    const GeometryContext& gctx, const cluster_t& cluster) const {
+  auto meas = cluster.measurement();
+  auto slink = std::visit([](const auto& x) { return x.sourceLink(); }, meas);
+  const auto geoId = slink.geometryId();
+  const Acts::Surface* surface = m_config.trackingGeometry->findSurface(geoId);
+  auto [localPos, localCov] = std::visit(
+      [](const auto& measurement) {
+        auto expander = measurement.expander();
+        Acts::BoundVector par = expander * measurement.parameters();
+        Acts::BoundSymMatrix cov =
+            expander * measurement.covariance() * expander.transpose();
+        // extract local position
+        Acts::Vector2 lpar(par[Acts::eBoundLoc0], par[Acts::eBoundLoc1]);
+        // extract local position covariance.
+        Acts::SymMatrix2 lcov =
+            cov.block<2, 2>(Acts::eBoundLoc0, Acts::eBoundLoc0);
+        return std::make_pair(lpar, lcov);
+      },
+      meas);
+
+  // transform local position to global coordinates
+  Acts::Vector3 globalFakeMom(1, 1, 1);
+
+  Acts::Vector3 globalPos =
+      surface->localToGlobal(gctx, localPos, globalFakeMom);
+  Acts::RotationMatrix3 rotLocalToGlobal =
+      surface->referenceFrame(gctx, globalPos, globalFakeMom);
+
+  // the space point requires only the variance of the transverse and
+  // longitudinal position. reduce computations by transforming the
+  // covariance directly from local to rho/z.
+  //
+  // compute Jacobian from global coordinates to rho/z
+  //
+  //         rho = sqrt(x² + y²)
+  // drho/d{x,y} = (1 / sqrt(x² + y²)) * 2 * {x,y}
+  //             = 2 * {x,y} / r
+  //       dz/dz = 1 (duuh!)
+  //
+  auto x = globalPos[Acts::ePos0];
+  auto y = globalPos[Acts::ePos1];
+  auto scale = 2 / std::hypot(x, y);
+  Acts::ActsMatrix<2, 3> jacXyzToRhoZ = Acts::ActsMatrix<2, 3>::Zero();
+  jacXyzToRhoZ(0, Acts::ePos0) = scale * x;
+  jacXyzToRhoZ(0, Acts::ePos1) = scale * y;
+  jacXyzToRhoZ(1, Acts::ePos2) = 1;
+  // compute Jacobian from local coordinates to rho/z
+  Acts::ActsMatrix<2, 2> jac =
+      jacXyzToRhoZ * rotLocalToGlobal.block<3, 2>(Acts::ePos0, Acts::ePos0);
+  // compute rho/z variance
+  Acts::ActsVector<2> var = (jac * localCov * jac.transpose()).diagonal();
+
+  auto gcov = Acts::Vector2(var[0], var[1]);
+  return std::make_pair(globalPos, gcov);
+}
+
+template <typename spacepoint_t, typename cluster_t>
+void Acts::SingleHitSpacePointBuilder<spacepoint_t, cluster_t>::
+    calculateSpacePoints(const GeometryContext& gctx,
+                         const std::vector<cluster_t>& clusters,
+                         std::vector<spacepoint_t>& spacePointStorage) const {
+  // Set the space point for all stored hits
+  for (const auto& clus : clusters) {
+    auto measurement = clus.measurement();
+    auto [gPos, gCov] = globalCoords(gctx, clus);
+    auto slink =
+        std::visit([](const auto& x) { return x.sourceLink(); }, measurement);
+
+    auto sp = spacepoint_t(gPos, gCov[0], gCov[1], {slink.index()});
```

Comment:
This assumes the space point class constructors. How should we handle ad-hoc constructors from the customers? Would this imply we'll have to create a space point collection with an acts-class and then convert (or copy) that to any customer-defined space point class?

Reply:
maybe this is more a question for @paulgessinger and @asalzburger since is more a general question also affecting other pieces of track reconstruction?

Reply:
Hm, good question. I guess this kind of depends on whether we keep the separation between internal and external spacepoints.

Reply:
There's a similar pattern with `std::function` used in the vertexing to convert input tracks, if I remember correctly. One concern is of course the potential virtual call overhead, although that might not be dominant here (and therefore not matter).